pageserver/src/disk_usage_eviction_task.rs

```diff
@@ -469,106 +443,160 @@ pub async fn disk_usage_eviction_task_iteration_impl<U: Usage>(
     cancel: &CancellationToken,
-) -> ControlFlow<()> {
-    debug!("begin");
-
-    let tenant = match tenant::mgr::get_tenant(tenant_id, true).await {
-        Ok(tenant) => tenant,
-        Err(e) => {
-            // this can happen if tenant has lifecycle transition after we fetched it
-            debug!("failed to get tenant: {e:#}");
-            return ControlFlow::Continue(());
-        }
-    };
-
-    if cancel.is_cancelled() {
-        return ControlFlow::Break(());
-    }
+) -> anyhow::Result<EvictionCandidates> {
+    // get a snapshot of the list of tenants
+    let tenants = tenant::mgr::list_tenants()
+        .await
+        .context("get list of tenants")?;
 
-    // If one of the timelines becomes `!is_active()` during the iteration,
-    // for example because we're shutting down, then `max_layer_size` can be too small.
-    // That's OK. This code only runs under a disk pressure situation, and being
-    // a little unfair to tenants during shutdown in such a situation is tolerable.
-    let mut max_layer_size = 0;
-    for tl in tenant.list_timelines() {
-        if !tl.is_active() {
-            continue;
-        }
-        let info = tl.get_local_layers_for_disk_usage_eviction();
-        debug!(timeline_id=%tl.timeline_id, "timeline resident layers count: {}", info.resident_layers.len());
-        scratch.extend(
-            info.resident_layers
-                .into_iter()
-                .map(|layer_infos| (tl.clone(), layer_infos)),
-        );
-        max_layer_size = max_layer_size.max(info.max_layer_size.unwrap_or(0));
+    let mut candidates = Vec::new();
 
+    for (tenant_id, _state) in &tenants {
         if cancel.is_cancelled() {
-            return ControlFlow::Break(());
+            return Ok(EvictionCandidates::Cancelled);
         }
-    }
+        let tenant = match tenant::mgr::get_tenant(*tenant_id, true).await {
+            Ok(tenant) => tenant,
+            Err(e) => {
+                // this can happen if tenant has lifecycle transition after we fetched it
+                debug!("failed to get tenant: {e:#}");
+                continue;
+            }
+        };
+
+        // collect layers from all timelines in this tenant
+        //
+        // If one of the timelines becomes `!is_active()` during the iteration,
+        // for example because we're shutting down, then `max_layer_size` can be too small.
+        // That's OK. This code only runs under a disk pressure situation, and being
+        // a little unfair to tenants during shutdown in such a situation is tolerable.
+        let mut tenant_candidates = Vec::new();
+        let mut max_layer_size = 0;
+        for tl in tenant.list_timelines() {
+            if !tl.is_active() {
+                continue;
+            }
+            let info = tl.get_local_layers_for_disk_usage_eviction();
+            debug!(timeline_id=%tl.timeline_id, "timeline resident layers count: {}", info.resident_layers.len());
```

Comment:
Timeline ids are very much unique but still thinking we could include tenant_id here for completeness and being alined with all other logging.